java-checks/src/main/java/org/sonar/java/checks/tests/MockitoAnnotatedObjectsShouldBeInitializedCheck.java

```diff
@@ -0,0 +1,177 @@
+
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import org.sonar.check.Rule;
+import org.sonar.plugins.java.api.IssuableSubscriptionVisitor;
+import org.sonar.plugins.java.api.semantic.MethodMatchers;
+import org.sonar.plugins.java.api.tree.AnnotationTree;
+import org.sonar.plugins.java.api.tree.BaseTreeVisitor;
+import org.sonar.plugins.java.api.tree.ClassTree;
+import org.sonar.plugins.java.api.tree.ExpressionTree;
+import org.sonar.plugins.java.api.tree.MethodInvocationTree;
+import org.sonar.plugins.java.api.tree.MethodTree;
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.plugins.java.api.tree.VariableTree;
+
+@Rule(key = "S5979")
+public class MockitoAnnotatedObjectsShouldBeInitializedCheck extends IssuableSubscriptionVisitor {
+  private static final List<String> TARGET_ANNOTATIONS = Arrays.asList(
+    "Captor",
+    "InjectMocks",
+    "Mock",
+    "Spy"
+  );
+
+  private static final List<String> EXPECTED_CLASS_ANNOTATIONS = Arrays.asList(
+    "ExtendWith",
+    "RunWith"
+  );
+
+  private static final List<String> BEFORE_ANNOTATIONS = Arrays.asList(
+    "Before",
+    "BeforeEach"
+  );
+
+  private static final MethodMatchers MOCKITO_JUNIT_RULE = MethodMatchers.create()
+    .ofAnyType()
+    .names("rule")
+    .addWithoutParametersMatcher()
+    .build();
+
+  private static final String MESSAGE = "Initialize mocks before using them.";
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    return Collections.singletonList(Tree.Kind.CLASS);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    ClassTree testClass = (ClassTree) tree;
+    List<VariableTree> mocksToInitialize = testClass.members().stream()
+      .filter(MockitoAnnotatedObjectsShouldBeInitializedCheck::isFieldWithTargetAnnotation)
+      .map(VariableTree.class::cast)
+      .collect(Collectors.toList());
+    if (!mocksToInitialize.isEmpty() && !mocksAreProperlyInitialized(testClass)) {
+      AnnotationTree firstAnnotation = mocksToInitialize.get(0).modifiers().annotations().get(0);
+      reportIssue(firstAnnotation, MESSAGE);
+    }
+  }
+
+  private static boolean isFieldWithTargetAnnotation(Tree tree) {
+    if (!tree.is(Tree.Kind.VARIABLE)) {
+      return false;
+    }
+    VariableTree field = (VariableTree) tree;
+    List<AnnotationTree> annotations = field.modifiers().annotations();
+    return !getAnnotations(annotations, TARGET_ANNOTATIONS).isEmpty();
+  }
+
+  private static boolean mocksAreProperlyInitialized(ClassTree testClass) {
+    return isClassProperlyAnnotated(testClass) ||
+      isMockitoJUnitRuleInvoked(testClass) ||
+      areMocksInitializedInSetup(testClass);
+  }
+
+  public static boolean isClassProperlyAnnotated(ClassTree clazz) {
+    List<AnnotationTree> annotations = clazz.modifiers().annotations();
+    return !getAnnotations(annotations, EXPECTED_CLASS_ANNOTATIONS).isEmpty();
```

Comment:
We don't test if it's `MockitoJUnitRunner` or `MockitoExtension`. It's only false-negatives, but we should add a test to make it explicit by adding a false-negative with a class annotated with for example `@RunWith(Parameterized.class)`